feat(OcppAuth): Updates to id_token_info for token in use on connector

mlitre opened the pull request:
We can now update id token info for a specific connector in case we receive updated info in a transaction event or start transaction message from OCPP. This allows to properly support new edges cases for the parent id tag. This is required for OCPP certification.

## Describe your changes

## Issue ticket number and link

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] I have made corresponding changes to the documentation
- [x] I read the [contribution documentation](https://github.com/EVerest/EVerest/blob/main/CONTRIBUTING.md) and made sure that my changes meet its requirements

Pietfried commented on modules/Auth/lib/AuthHandler.cpp:
```diff
@@ -124,6 +124,20 @@ TokenHandlingResult AuthHandler::on_token(const ProvidedIdToken& provided_token)
+        this->evses.at(connector_id)->identifier->authorization_status =
+            validation_result_update.validation_result.authorization_status;
+        this->evses.at(connector_id)->identifier->parent_id_token =
+            validation_result_update.validation_result.parent_id_token;
```
You can maybe add a comment here that other parts of the `ValidationResult` are neglected for now

Pietfried commented on interfaces/auth_token_validator.yaml:
```diff
@@ -13,5 +13,10 @@ cmds:
       $ref: /authorization#/ValidationResult
+vars:
+  validate_result_update:
+    description: Updated validation result for a given token due to some changes.
```
You could add in the description that it was added because OCPP can do an update in `StartTransaction.req` or `TransactionEvent.conf` . It makes it a little easier to argue about it later if we need to

Pietfried commented on modules/Auth/lib/AuthHandler.cpp:
```diff
@@ -124,6 +124,20 @@ TokenHandlingResult AuthHandler::on_token(const ProvidedIdToken& provided_token)
 }
 
+void AuthHandler::handle_token_validation_result_update(const ValidationResultUpdate& validation_result_update) {
+    auto connector_id = validation_result_update.connector_id;
```
This should be protected using the `event_mutex` since `this->evses` is accessed and updated.  
```suggestion
    std::unique_lock<std::mutex> lk(this->event_mutex);
    auto connector_id = validation_result_update.connector_id;
```

Pietfried commented on modules/Auth/Auth.cpp:
```diff
@@ -24,6 +24,14 @@ void Auth::init() {
     }
+    for (const auto& token_validator : this->r_token_validator) {
+        token_validator->subscribe_validate_result_update([this](ValidationResultUpdate validation_result_update) {
+            std::thread t([this, validation_result_update]() {
```
You dont have to run this in a new thread. It's done for `on_token` because `on_token` potentially blocks and/or takes some time to process the request. In order to handle multiple requests from the same provider, it's running it a seperate thread.

Since `handle_token_validation_result_update` can be processed and return without any waiting, its not required to run it in a separate thread.

mlitre replied:
Done

Pietfried reviewed (changes requested):
Looks pretty good overall, I added some remarks inline. It would be nice to have 1-2 OCPP e2e tests for this

mlitre commented:
> Looks pretty good overall, I added some remarks inline. It would be nice to have 1-2 OCPP e2e tests for this

I added a test for OCPP 1.6 and OCPP 2.0.1

Pietfried commented on modules/Auth/lib/AuthHandler.cpp:
```diff
@@ -124,6 +124,22 @@ TokenHandlingResult AuthHandler::on_token(const ProvidedIdToken& provided_token)
+        this->evses.at(connector_id)->identifier->authorization_status =
+            validation_result_update.validation_result.authorization_status;
+        this->evses.at(connector_id)->identifier->parent_id_token =
+            validation_result_update.validation_result.parent_id_token;
```
I think we should also call `publish_token_validation_status_callback` with the new status and updated provided id token here

mlitre replied:
Done

Pietfried commented on tests/ocpp_tests/test_sets/ocpp201/transactions.py:
```diff
@@ -453,3 +454,140 @@ async def test_two_parallel_transactions(
+    # stop charging session
+    test_controller.plug_out()
+
+    await asyncio.sleep(10)
```
Can we drop this?

mlitre replied:
Ah yes we probably can, it came from the TC that I copied
